phoenix-core/src/it/java/org/apache/phoenix/query/MaxConcurrentConnectionsIT.java

```diff
@@ -0,0 +1,152 @@
+    }
+
+    //Have to shutdown our special delayed region server
+    @AfterClass
+    public static void tearDown() throws Exception {
+        hbaseTestUtil.shutdownMiniCluster();
+    }
+
+    @Test
+    public void testCannotCreateOverMaximumConnections() throws Exception {
+        int maxConnections = 3;
+        Properties props = new Properties();
+        props.setProperty(CLIENT_CONNECTION_MAX_ALLOWED_CONNECTIONS,String.valueOf(maxConnections));
+        props.setProperty(RENEW_LEASE_ENABLED,String.valueOf(false));
+        assertEquals(0,GLOBAL_OPEN_PHOENIX_CONNECTIONS.getMetric().getValue());
+
+        List<Connection> connections = new ArrayList<>();
+        String url = getUniqueUrl();
+
+        try{
+            for(int i = 0; i < maxConnections; i++) {
+                connections.add(DriverManager.getConnection(url, props));
+            }
+
+            try {
+                DriverManager.getConnection(url, props);
+                fail();
+            } catch(SQLException e) {
+                assertEquals(NEW_CONNECTION_THROTTLED.getErrorCode(),e.getErrorCode());
+                assertEquals(NEW_CONNECTION_THROTTLED.getSQLState(),e.getSQLState());
+            }
+
+        } finally {
+            for(Connection conn : connections) {
+                conn.close();
+            }
+        }
+    }
+
+    /**
+     * This tests the delete path which creates a internal phoenix connection per region
+     * @throws Exception
+     */
+    @Test
+    public void testDeleteRuntimeFailureClosesConnections() throws Exception {
+        String tableName = generateUniqueName();
+
+        //table with lots of regions
+        String ddl = "create table " + tableName +  "  (i integer not null primary key, j integer) SALT_BUCKETS=256 ";
+
+        Properties props = new Properties();
+        props.setProperty(QueryServices.ENABLE_SERVER_SIDE_DELETE_MUTATIONS,
+                String.valueOf(false));
+        props.setProperty(CLIENT_CONNECTION_MAX_ALLOWED_CONNECTIONS,String.valueOf(10));
+        props.setProperty(INTERNAL_CONNECTION_MAX_ALLOWED_CONNECTIONS,String.valueOf(10));
+
+        String deleteStmt = "DELETE FROM " + tableName + " WHERE 20 = j";
+
+        assertEquals(0,GLOBAL_OPEN_PHOENIX_CONNECTIONS.getMetric().getValue());
+        assertEquals(0,GLOBAL_OPEN_INTERNAL_PHOENIX_CONNECTIONS.getMetric().getValue());
+        Connection conn = null;
+        try{
+            conn = DriverManager.getConnection(getUniqueUrl(), props);
+            try(Statement statement = conn.createStatement()) {
+                statement.execute(ddl);
+            }
+            //Enable delay for the delete
+            DelayedRegionServer.setDelayEnabled(true);
+            try(Statement statement = conn.createStatement()) {
+                statement.execute(deleteStmt);
+            }
+            fail();
+        } catch (SQLException e) {
+            assertEquals(NEW_CONNECTION_THROTTLED.getErrorCode(),e.getErrorCode());
```

Comment:
So here, the client opens just 1 connection and the delete path creates 1 cloned conn per region and so this leads to client throttling? I thought internal conns wouldn't get counted in the client conn throttling?

Reply:
This hits the internal throttling defined as 10.

Reply:
Ok, as I think I've mentioned elsewhere, we should have a separate exception for internal vs external conn throttling.

Reply:
Done